### PR TITLE
Ensure Job sync invocations are batched by 1s periods

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

It is possible that `syncJob` invocations are triggered without queuing, see: https://github.com/kubernetes/kubernetes/blob/72a3990728b2a8979effb37b9800beb3117349f6/pkg/controller/job/job_controller.go#L443. 

This is the scenario:
- `pod_1` completes triggering `syncJob`, resulting in a delayed *Job status update* (to increment the counter for `pod_1`)
- job status update completes, requeing another `syncJob` without delay
- `pod_2` completes 
- the last non-delayed `syncJob` starts, and results in Job status update (to increment the counter for `pod_2`)
- job status update completes, requeuing another `syncJob` without delay
- `pod_3` completes
-  the last non-delayed `syncJob` starts, and results in Job status update (to increment the counter for `pod_3`)
...

If we have enough pods terminating in between the consecutive `syncJob`'s, then the invocations may create a chain of immediate invocations. 

While the scenario is unlikely, ensuring the syncJobs are queued with delay will make it easier to reason about the queuing mechanism.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Ensure Job status updates are batched by 1s. This fixes an unlikely scenario when a sequence of immediately 
completing pods could trigger a sequence of non-batched Job status updates.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -440,7 +440,14 @@ func (jm *Controller) updateJob(old, cur interface{}) {
 	if err != nil {
 		return
 	}
-	jm.enqueueController(curJob, true)
+	if curJob.Generation == oldJob.Generation {
+		// Delay the Job sync when no generation change to batch Job status updates,
+		// typically triggered by pod events.
+		jm.enqueueControllerDelayed(curJob, true)
+	} else {
+		// Trigger immediate sync when spec is changed.
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:448]:
```suggestion
		// Trigger immediate sync when spec changes.
```

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve
/hold
to verify how bad could this be. This depends on the latency on a typical mutating call to the apiserver (the status update).

**Comment by mimowo** [comment]
> /hold to verify how bad could this be. This depends on the latency on a typical mutating call to the apiserver (the status update).

Would you like to carry an actual test, or an estimation based on some assumptions?

**Comment by alculquicondor** [comment]
just estimation. I'm tempted to mark this as `cleanup` instead of `bug`.

**Comment by mimowo** [comment]
Withdrawn for now from the delay on conflict. There is a unit test checking it is requeued immediately, that started to fail. Maybe this scenario should be considered separately.

**Review by soltysh — COMMENTED** [review]
The changes here look good, how about an integration test which exercises this path? Not sure if it's possible to trigger this units, but that's an option too.

**Comment by mimowo** [comment]
> The changes here look good, how about an integration test which exercises this path? Not sure if it's possible to trigger this units, but that's an option too.

I think unit tests might actually be simpler to add as we have a fake clock, we could just check that the `syncJob` isn't requeued immediately, but it is when we shift the clock. Then, we could test that Job spec update requeues `syncJob` immediately. This is important, for example, when changing `spec.suspend` true to false. I will try to add some along the lines.

As for integration tests,  I'm less sure how to design them.

File: pkg/controller/job/job_controller_test.go [hunk 2]
```diff
@@ -1894,10 +1894,6 @@ func TestPastDeadlineJobFinished(t *testing.T) {
 				t.Errorf("Could not create Job: %v", err)
 			}
 
-			if err := sharedInformerFactory.Batch().V1().Jobs().Informer().GetIndexer().Add(job); err != nil {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:1897]: :thinking: the unit test didn't used to work without these calls to the Indexer. Maybe the test client changed?

**Review by alculquicondor — COMMENTED** [review]
+1 on a unit test.

@soltysh do you think we should backport this?

Otherwise I would like to see a cleanup here: a better name for `immediate` or separate functions for delayed or not.

I think in the past we have tried to preserve the code that works, but I believe it's time to clean up.

File: pkg/controller/job/job_controller_test.go (hunk 2 above)
Comment (mimowo) [reply pkg/controller/job/job_controller_test.go:1897]: IIRC these lines were added in an attempt to get rid of flakes before we understood the underlying cause for the flakes: https://github.com/kubernetes/kubernetes/pull/114944#discussion_r1069179801. 

To check that, I have checked out the initial commits ([commit1](https://github.com/kubernetes/kubernetes/commit/4c1b95ddfafee38a8acd654aaec6e00829b3908b) and [commit2](https://github.com/kubernetes/kubernetes/commit/eea7dca085af3b32199b2d4210d229c2dd624791)) introducing these lines, removed it and the tests passed (but wouldn't pass consistently).

**Comment by mimowo** [comment]
Unit test added. PTAL.

**Comment by mimowo** [comment]
Reverted the namings are they will anyway be revisited when doing https://github.com/kubernetes/kubernetes/issues/118527

**Review by alculquicondor — COMMENTED** [review]
/hold cancel
/lgtm

**Pull request merged**